### DISCUSSION TITLE
SignTool Issue

### DISCUSSION THREAD

**Discussion by tstewart65** [cake-build/cake#4404] [Q&A]
The SignTool is missing the /kc and /csp parameters.  These are used for Private Key selection.  I have tried to use the ArgumentCustomization to add these arguments but it does not work.  The issue is the arguments are added to the end of the string which puts them after the list of files to be signed and they are treated as files.

Is there a way around this or another method for signing that can use this parameters?

**Comment by gep13** [comment]
@tstewart65 can you provide an example of the customization that you are trying to do?

While the typical examples of usage of ArgumentCustomization is to use the `Append` method, there is also a `Prepend` method.  There is even a `Clear` method to completely remove everything, and then add what is required back in again.

https://cakebuild.net/api/Cake.Core.IO/ProcessArgumentBuilder/

**Comment by tstewart65** [comment]
The customization I was trying to do is to add the /kc and /csp parameters that I need for signing since they are not supported in SignToolSignSettings.  These parameters are required to allow signing using a DigiCert USB Token which has the private key stored on it.  You can no longer obtain a code signing certificate that is not stored on an HSM.  The old certificates you could store in the local certificate manager which is no longer allowed.

The Append method does not work since it appends the arguments after the file list, and they are not recognized.  I did go back and try the Prepend method, but this also does not work since it places those arguments before the sign command which needs to be the first thing after signtool.exe.  The clear method would work but then I have to go and add all of the required arguments which is not ideal when it could be fixed by adding support for these two arguments.
